ksql-engine/src/main/java/io/confluent/ksql/planner/plan/FlatMapNode.java

```diff
@@ -0,0 +1,198 @@
+  }
+
+  @Override
+  protected int getPartitions(final KafkaTopicClient kafkaTopicClient) {
+    return source.getPartitions(kafkaTopicClient);
+  }
+
+  @Override
+  public SchemaKStream<?> buildStream(final KsqlQueryBuilder builder) {
+
+    final QueryContext.Stacker contextStacker = builder.buildNodeContext(getId().toString());
+
+    return getSource().buildStream(builder).flatMap(
+        outputSchema,
+        analysis.getTableFunctions().get(0),
+        contextStacker
+    );
+  }
+
+  private LogicalSchema buildLogicalSchema(final LogicalSchema inputSchema) {
+    final LogicalSchema.Builder schemaBuilder = LogicalSchema.builder();
+    final List<Column> cols = inputSchema.value();
+
+    // We copy all the original columns to the output schema
+    schemaBuilder.keyColumns(inputSchema.key());
+    for (Column col : cols) {
+      schemaBuilder.valueColumn(col);
+    }
+
+    final ConnectToSqlTypeConverter converter = SchemaConverters.connectToSqlConverter();
+
+    // And add new columns representing the exploded values at the end
+    for (int i = 0; i < analysis.getTableFunctions().size(); i++) {
+      final KsqlTableFunction tableFunction =
+          UdtfUtil.resolveTableFunction(functionRegistry,
+              analysis.getTableFunctions().get(i), inputSchema
+          );
+      final ColumnName colName = ColumnName.generatedColumnName(i);
+      final SqlType fieldType = converter.toSqlType(tableFunction.getReturnType());
+      schemaBuilder.valueColumn(colName, fieldType);
+    }
+
+    return schemaBuilder.build();
+  }
+
+  private List<SelectExpression> buildFinalSelectExpressions() {
+    final TableFunctionExpressionRewriter tableFunctionExpressionRewriter =
+        new TableFunctionExpressionRewriter();
+    final List<SelectExpression> selectExpressions = new ArrayList<>();
+    for (final SelectExpression select : analysis.getSelectExpressions()) {
+      final Expression exp = select.getExpression();
+      selectExpressions.add(
+          SelectExpression.of(
+              select.getAlias(),
+              ExpressionTreeRewriter.rewriteWith(
+                  tableFunctionExpressionRewriter::process, exp)
+          ));
+    }
+    return selectExpressions;
+  }
+
+  private class TableFunctionExpressionRewriter
+      extends VisitParentExpressionVisitor<Optional<Expression>, Context<Void>> {
+
+    private int variableIndex = 0;
+
+    TableFunctionExpressionRewriter() {
+      super(Optional.empty());
+    }
+
+    @Override
+    public Optional<Expression> visitFunctionCall(
+        final FunctionCall node,
+        final Context<Void> context
+    ) {
+      final String functionName = node.getName().name();
+      if (functionRegistry.isTableFunction(functionName)) {
+        final ColumnName varName = ColumnName.generatedColumnName(variableIndex);
+        variableIndex++;
+        return Optional.of(
+            new ColumnReferenceExp(node.getLocation(), ColumnRef.of(Optional.empty(), varName)));
```

Comment:
I think we should be building the selects and the logical schema in the same pass instead of in two passes and then hope that the variableIndex lines up properly with the colName generated above in `buildLogicalSchema`. This can be done in a future PR (preferably the same one I suggest above about introducing a `ProjectSchema`).

Reply:
https://github.com/confluentinc/ksql/issues/3659